File: speech-to-text/file-transcription/javascript/transcribe.js
#!/usr/bin/env node
/**
 * Smallest AI Speech-to-Text - File Transcription
 *
 * Transcribe audio files with advanced features like word timestamps,
 * speaker diarization, and emotion detection.
 *
 * Usage: node transcribe.js <audio_file>
 *
 * Output:
 * - Command line response with feature outputs
 * - {filename}_transcript.txt - Plain text transcription
 * - {filename}_result.json - Full API response
 */

const fs = require("fs");
const path = require("path");

const API_URL = "https://waves-api.smallest.ai/api/v1/pulse/get_text";
const OUTPUT_DIR = ".";

// The following are all the features supported by the POST endpoint (Pre-Recorded API)
const LANGUAGE = "en"; // Use ISO 639-1 codes or "multi" for auto-detect
const WORD_TIMESTAMPS = true;
const DIARIZE = true;
const AGE_DETECTION = true;
const GENDER_DETECTION = true;
const EMOTION_DETECTION = true;

async function transcribe(audioFile, apiKey) {
  const audioData = fs.readFileSync(audioFile);

  const params = new URLSearchParams({
    language: LANGUAGE,
    word_timestamps: WORD_TIMESTAMPS,
    diarize: DIARIZE,
    age_detection: AGE_DETECTION,
    gender_detection: GENDER_DETECTION,
    emotion_detection: EMOTION_DETECTION,
  });

  const response = await fetch(`${API_URL}?${params}`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${apiKey}`,
      "Content-Type": "application/octet-stream",
    },
    body: audioData,
  });

  if (!response.ok) {
    throw new Error(`API request failed with status ${response.status}: ${await response.text()}`);
  }

  return response.json();
}

// This function is designed to process feature outputs for all the features supported
// by the POST endpoint (Pre-Recorded API)
function processResponse(result, audioPath) {
  if (result.status !== "success") {
    console.error("Error: Transcription failed");
    console.error(result);
    process.exit(1);
  }

  console.log("\n" + "=".repeat(60));
  console.log("TRANSCRIPTION");
  console.log("=".repeat(60));
  console.log(result.transcription || "");

  // Speaker info
  if (result.age || result.gender || result.emotions) {
    console.log("\n" + "-".repeat(60));
    console.log("SPEAKER INFO");
    console.log("-".repeat(60));
    if (result.gender) {
      console.log(`Gender: ${result.gender}`);
    }
    if (result.age) {
      console.log(`Age: ${result.age}`);
    }
    if (result.emotions) {
      console.log("Emotions:");
      const sortedEmotions = Object.entries(result.emotions).sort((a, b) => b[1] - a[1]);
      for (const [emotion, score] of sortedEmotions) {
        console.log(`  ${emotion}: ${score.toFixed(4)}`);
      }
    }
  }

  // Utterances (speaker diarization)
  if (result.utterances && result.utterances.length > 0) {
    console.log("\n" + "-".repeat(60));
    console.log("UTTERANCES");
    console.log("-".repeat(60));
    for (const utt of result.utterances) {
      const speaker = utt.speaker || "unknown";
      const start = utt.start || 0;
      const end = utt.end || 0;
      const text = utt.text || "";
      console.log(`[${start.toFixed(2)}s - ${end.toFixed(2)}s] ${speaker}: ${text}`);
    }
  }

  // Word timestamps
  if (result.words && result.words.length > 0) {
    console.log("\n" + "-".repeat(60));
    console.log("WORD TIMESTAMPS");
    console.log("-".repeat(60));
    for (const word of result.words) {
      const start = word.start || 0;
      const end = word.end || 0;
      const text = word.word || "";
      const speaker = word.speaker || "";
      if (speaker) {
        console.log(`[${start.toFixed(2)}s - ${end.toFixed(2)}s] ${speaker}: ${text}`);
      } else {
        console.log(`[${start.toFixed(2)}s - ${end.toFixed(2)}s] ${text}`);
      }
    }
  }

  console.log("\n" + "=".repeat(60));

  if (!fs.existsSync(OUTPUT_DIR)) {
    fs.mkdirSync(OUTPUT_DIR, { recursive: true });
  }

  const textPath = path.join(OUTPUT_DIR, `${audioPath.name}_transcript.txt`);
  fs.writeFileSync(textPath, result.transcription || "", "utf-8");
  console.log(`Saved: ${textPath}`);

  const jsonPath = path.join(OUTPUT_DIR, `${audioPath.name}_result.json`);
  fs.writeFileSync(jsonPath, JSON.stringify(result, null, 2), "utf-8");
  console.log(`Saved: ${jsonPath}`);

  console.log("\nDone!");
}

async function main() {
  const audioFile = process.argv[2];

  if (!audioFile) {
    console.log("Usage: node transcribe.js <audio_file>");
    process.exit(1);
  }

  const apiKey = process.env.SMALLEST_API_KEY;

  if (!apiKey) {
    console.error("Error: SMALLEST_API_KEY environment variable not set");
    process.exit(1);
  }

  if (!fs.existsSync(audioFile)) {
    console.error(`Error: File not found: ${audioFile}`);
    process.exit(1);
  }

  const audioPath = path.parse(audioFile);
  console.log(`Reading: ${audioPath.base}`);
  console.log(`Transcribing with language: ${LANGUAGE}`);

  const result = await transcribe(audioFile, apiKey);
  processResponse(result, audioPath);
}

main();

File: speech-to-text/file-transcription/python/transcribe.py
#!/usr/bin/env python3
"""
Smallest AI Speech-to-Text - File Transcription

Transcribe audio files with advanced features like word timestamps,
speaker diarization, and emotion detection.

Usage: python transcribe.py <audio_file>

Output:
- Command line response with feature outputs
- {filename}_transcript.txt - Plain text transcription
- {filename}_result.json - Full API response
"""

import json
import os
import sys
from pathlib import Path

import requests

API_URL = "https://waves-api.smallest.ai/api/v1/pulse/get_text"
OUTPUT_DIR = "."

# The following are all the features supported by the POST endpoint (Pre-Recorded API)
LANGUAGE = "en"  # Use ISO 639-1 codes or "multi" for auto-detect
WORD_TIMESTAMPS = False
DIARIZE = False
AGE_DETECTION = False
GENDER_DETECTION = False
EMOTION_DETECTION = False


def transcribe(audio_file: str, api_key: str) -> dict:
    with open(audio_file, "rb") as f:
        audio_data = f.read()

    response = requests.post(
        API_URL,
        headers={
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/octet-stream",
        },
        params={
            "language": LANGUAGE,
            "word_timestamps": str(WORD_TIMESTAMPS).lower(),
            "diarize": str(DIARIZE).lower(),
            "age_detection": str(AGE_DETECTION).lower(),
            "gender_detection": str(GENDER_DETECTION).lower(),
            "emotion_detection": str(EMOTION_DETECTION).lower(),
        },
        data=audio_data,
        timeout=300,
    )

    if response.status_code != 200:
        raise Exception(f"API request failed with status {response.status_code}: {response.text}")

    return response.json()

# This function is designed to process feature outputs for all the features supported
# by the POST endpoint (Pre-Recorded API)
def process_response(result: dict, audio_path: Path):
    if result.get("status") != "success":
        print(f"Error: Transcription failed")
        print(result)
        sys.exit(1)

    print("\n" + "=" * 60)
    print("TRANSCRIPTION")
    print("=" * 60)
    print(result.get("transcription", ""))

    # Speaker info
    if result.get("age") or result.get("gender") or result.get("emotions"):
        print("\n" + "-" * 60)
        print("SPEAKER INFO")
        print("-" * 60)
        if result.get("gender"):
            print(f"Gender: {result['gender']}")
        if result.get("age"):
            print(f"Age: {result['age']}")
        if result.get("emotions"):
            print("Emotions:")
            emotions = result["emotions"]
            sorted_emotions = sorted(emotions.items(), key=lambda x: x[1], reverse=True)
            for emotion, score in sorted_emotions:
                print(f"  {emotion}: {score:.4f}")

    # Utterances (speaker diarization)
    if result.get("utterances"):
        print("\n" + "-" * 60)
        print("UTTERANCES")
        print("-" * 60)
        for utt in result["utterances"]:
            speaker = utt.get("speaker", "")
            start = utt.get("start", 0)
            end = utt.get("end", 0)
            text = utt.get("text", "")
            if speaker:
                print(f"[{start:.2f}s - {end:.2f}s] {speaker}: {text}")
            else:
                print(f"[{start:.2f}s - {end:.2f}s] {text}")

    # Word timestamps
    if result.get("words"):
        print("\n" + "-" * 60)
        print("WORD TIMESTAMPS")
        print("-" * 60)
        for word in result["words"]:
            start = word.get("start", 0)
            end = word.get("end", 0)
            text = word.get("word", "")
            speaker = word.get("speaker", "")
            if speaker:
                print(f"[{start:.2f}s - {end:.2f}s] {speaker}: {text}")
            else:
                print(f"[{start:.2f}s - {end:.2f}s] {text}")

    print("\n" + "=" * 60)

    output_dir = Path(OUTPUT_DIR)
    output_dir.mkdir(parents=True, exist_ok=True)

    text_path = output_dir / f"{audio_path.stem}_transcript.txt"
    with open(text_path, "w", encoding="utf-8") as f:
        f.write(result.get("transcription", ""))
    print(f"Saved: {text_path}")

    json_path = output_dir / f"{audio_path.stem}_result.json"
    with open(json_path, "w", encoding="utf-8") as f:
        json.dump(result, f, indent=2, ensure_ascii=False)
    print(f"Saved: {json_path}")

    print("\nDone!")


def main():
    if len(sys.argv) < 2:
        print("Usage: python transcribe.py <audio_file>")
        sys.exit(1)

    audio_file = sys.argv[1]
    api_key = os.environ.get("SMALLEST_API_KEY")

    if not api_key:
        print("Error: SMALLEST_API_KEY environment variable not set")
        sys.exit(1)

    audio_path = Path(audio_file)
    if not audio_path.exists():
        print(f"Error: File not found: {audio_file}")
        sys.exit(1)

    print(f"Reading: {audio_path.name}")
    print(f"Transcribing with language: {LANGUAGE}")

    result = transcribe(audio_file, api_key)
    process_response(result, audio_path)


if __name__ == "__main__":
    main()
